src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -748,3 +750,80 @@ impl RustcTargetData {
         self.target_config(kind).links_overrides.get(lib_name)
     }
 }
+
+/// Structure used to deal with Rustdoc fingerprinting
+#[derive(Debug, Serialize, Deserialize)]
+pub struct RustDocFingerprint {
+    rustc_verbose_version: String,
+}
+
+impl RustDocFingerprint {
+    /// Returns the version contained by this `RustDocFingerprint` instance.
+    fn version(&self) -> &str {
+        self.rustc_verbose_version.as_str()
+    }
+
+    /// Given the `doc` dir, returns the path to the rustdoc fingerprint file.
+    fn path_to_fingerprint(doc_dir: &Path) -> PathBuf {
+        doc_dir.to_path_buf().join(".rustdoc_fingerprint.json")
+    }
+
+    /// Write the `RustDocFingerprint` info into the fingerprint file.
+    pub fn write(&self, doc_dir: &Path) -> std::io::Result<()> {
+        crate::util::paths::create_dir_all(doc_dir)
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("{:?}", e)))?;
+        let rustdoc_fingerprint_file =
+            File::create(RustDocFingerprint::path_to_fingerprint(doc_dir))?;
+        // We write the actual `Rustc` version to it so that we just need to compile it straight
+        // since there's no `doc/` folder to remove.
+        serde_json::to_writer(&rustdoc_fingerprint_file, &self)
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("{:?}", e)))
+    }
+
+    /// This function checks whether the latest version of `Rustc` used to compile this
+    /// `Workspace`'s docs was the same as the one is currently being used in this `cargo doc`
+    /// call, returning `(Self, bool)` where the `bool` says if the `doc` folder was removed.
+    ///
+    /// In case it's not, it takes care of removig the `doc/` folder as well as overwriting
+    /// the rustdoc fingerprint info in order to guarantee that we won't end up with mixed
+    /// versions of the `js/html/css` files that `Rustc` autogenerates which do not have
+    /// any versioning.
+    pub fn check_rustdoc_fingerprint(
+        doc_dir: &Path,
+        rustc_verbose_ver: &str,
+    ) -> anyhow::Result<(Self, bool)> {
+        let actual_rustdoc_target_data = RustDocFingerprint {
+            rustc_verbose_version: rustc_verbose_ver.to_string(),
+        };
+
+        // Check wether `.rustdoc_fingerprint.json exists
+        match std::fs::read_to_string(Self::path_to_fingerprint(doc_dir)) {
```

Comment:
Generally in Cargo we try to avoid using `fs` functions directly.  The `util::paths` module contains several wrappers, such as `paths::read` to handle reading.